kernel/src/log_segment/mod.rs

```diff
@@ -0,0 +1,374 @@
+                metadata_opt = Metadata::try_new_from_data(batch.as_ref())?;
+            }
+            if protocol_opt.is_none() {
+                protocol_opt = Protocol::try_new_from_data(batch.as_ref())?;
+            }
+            if metadata_opt.is_some() && protocol_opt.is_some() {
+                // we've found both, we can stop
+                break;
+            }
+        }
+        match (metadata_opt, protocol_opt) {
+            (Some(m), Some(p)) => Ok((m, p)),
+            (None, Some(_)) => Err(Error::MissingMetadata),
+            (Some(_), None) => Err(Error::MissingProtocol),
+            (None, None) => Err(Error::MissingMetadataAndProtocol),
+        }
+    }
+
+    // Replay the commit log, projecting rows to only contain Protocol and Metadata action columns.
+    fn replay_for_metadata(
+        &self,
+        engine: &dyn Engine,
+    ) -> DeltaResult<impl Iterator<Item = DeltaResult<(Box<dyn EngineData>, bool)>> + Send> {
+        let schema = get_log_schema().project(&[PROTOCOL_NAME, METADATA_NAME])?;
+        // filter out log files that do not contain metadata or protocol information
+        static META_PREDICATE: LazyLock<Option<ExpressionRef>> = LazyLock::new(|| {
+            Some(Arc::new(Expression::or(
+                Expression::column([METADATA_NAME, "id"]).is_not_null(),
+                Expression::column([PROTOCOL_NAME, "minReaderVersion"]).is_not_null(),
+            )))
+        });
+        // read the same protocol and metadata schema for both commits and checkpoints
+        self.replay(engine, schema.clone(), schema, META_PREDICATE.clone())
+    }
+}
+
+/// Returns a fallible iterator of [`ParsedLogPath`] that are between the provided `start_version` (inclusive)
+/// and `end_version` (inclusive). [`ParsedLogPath`] may be a commit or a checkpoint.  If `start_version` is
+/// not specified, the files will begin from version number 0. If `end_version` is not specified, files up to
+/// the most recent version will be included.
+///
+/// Note: this calls [`FileSystemClient::list_from`] to get the list of log files.
+fn list_log_files(
+    fs_client: &dyn FileSystemClient,
+    log_root: &Url,
+    start_version: impl Into<Option<Version>>,
+    end_version: impl Into<Option<Version>>,
+) -> DeltaResult<impl Iterator<Item = DeltaResult<ParsedLogPath>>> {
+    let start_version = start_version.into().unwrap_or(0);
+    let end_version = end_version.into();
+
+    let version_prefix = format!("{:020}", start_version);
+    let start_from = log_root.join(&version_prefix)?;
+
+    Ok(fs_client
+        .list_from(&start_from)?
+        .map(|meta| ParsedLogPath::try_from(meta?))
+        // TODO this filters out .crc files etc which start with "." - how do we want to use these kind of files?
+        .filter_map_ok(identity)
+        .take_while(move |path_res| match path_res {
+            Ok(path) => end_version.is_none_or(|end_version| path.version <= end_version),
+            Err(_) => true,
```

Comment:
Rescuing https://github.com/delta-incubator/delta-kernel-rs/pull/495#discussion_r1843137629
> Hmm is it a good idea to be filtering out errors though?

This is a good question we should probably track as a separate discussion/issue:

The `ParsedLogPath` code recognizes four general classes of files:
1. "Known" versioned files, e.g. `Ok(Some(SinglePartCheckpoint))`
2. "Unknown" versioned files that should be ignored, `Ok(Some(Unknown))`
3. Non-versioned files that should be ignored, `Ok(None)`
4. Versioned files that resemble a known type but which fail to parse, `Err(_)` 

The [Delta spec](https://github.com/delta-io/delta/blob/master/PROTOCOL.md#protocol-evolution) doesn't specifically say what to do with unrecognized file types, but it is pretty clear that we should ignore unrecognized things in general:
> Since breaking changes must be accompanied by an increase in the protocol version recorded in a table or by the addition of a table feature, clients can assume that unrecognized actions, fields, and/or metadata domains are never required in order to correctly interpret the transaction log. Clients must ignore such unrecognized fields, and should not produce an error when reading a table that contains unrecognized fields.

... and some Delta table features like v2 checkpoints and log compaction rely on older clients to ignore unrecognized files/dirs.

So then the question is whether file names like these:
```
# hex instead of decimal
00000000deadbeef.commit.json  

# bogus part numbering
00000000000000000000.checkpoint.0000000010.0000000000.parquet 

# v2 checkpoint, as seen by a client that doesn't understand that feature
00000000000000000010.checkpoint.80a083e8-7026-4e79-81be-64bd76c43a11.json

# compacted log file, as seen by a client that doesn't understand that feature
00000000000000000004.00000000000000000006.compacted.json
```
... are simply "unrecognized files" or actual errors to care about. Your code is treating them like actual errors; @zachschuermann is suggesting to just filter them out.

Maybe we should take the middle ground and filter them out with a `warn!` so there's at least a possibility to know they happened?

Reply:
Note: The original code (before this PR) was propagating file name parsing errors, so we should probably keep that behavior for now and track this discussion separate follow-up?

Reply:
yea agree with the last comment -- let's keep behavior and file a follow up (and if we want to add a warn! for some other cases that's fine too)

Reply:
Thx for the context @scovich!! Agreed that it's something to address. No need to fail if we don't have to. I'll talk to @zachschuermann  and get an issue up